javascript/computer_science/project_hash_map.md

```diff
@@ -0,0 +1,227 @@
+
+To deal with this, our hash map class need to keep track of two new fields, the `capacity` and the `load factor`.
+
+- The `capacity` is the amount of buckets we currently have. Keeping track of this will let us know if our map has reached the threshold, the threshold will depend on our factor variable as well.
+
+- The `load factor` is a number that we can assign our hash map to at the start. It's the factor that will determine when is it a good time to grow our buckets, for example a load factor of `0.75` means our hash map will need to grow its buckets when the capacity reaches 75% full. Setting it too low will consume too much memory by having too many empty buckets, while setting it too high will allow our buckets to have collisions before we grow them. Usually a good balance of `0.75 to 1` is used.
+
+### Computation complexity
+
+A hash map is very efficient in its insertion, retrieving and removing operations. That is because we use array indexes to do those operations. A hash map has an average case complexity of `O(1)` for the following methods:
+- Insertion
+- Retrieving
+- Deletion
+
+Assuming we have a good hash map written. The worst case of those operations would be `O(n)` and that happens when we have all our data hashes to the same exact bucket.
+
+The Growth of our hash map has the complexity of `O(n)` at all times.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+#### Self restriction and limitation
+
+  Before we get started, we need to lay down some ground rules. As we all know JavaScript is a dynamic language the restriction we have in other languages does not exist in JavaScript for example if you create an array of size `16` to be your buckets size, nothing stopping you from storing items at index `500`. This defeats the purpose we are trying to demonstrate, so we need to put some self restriction to work around this.
+
+  1. When you create a function that takes a hash code value and returns the bucket index, use this snippet before the function returns to make sure it's accessing inbound array indexes only.
+
+      ```javascript
+      if (index < 0 || index >= buckets.length) {
+        throw new Error("Trying to access index out of bound");
+      }
+      ```
+
+  1. Inside the class DO NOT use Objects to store your key value pairs, And DO NOT use the array method `push` to insert your nodes into the buckets, always use indexes.
+
+#### Time to create a real hash map
+
+  1. Start by creating a HashMap class or factory function, up to you. Then proceed to create the following methods:
+
+  1. `hash` takes a value and produces a hash code with it. There are a lot of thought that goes into creating the most optimal hashing function, it is up to you if you want to do a research to find a good hashing function but do not dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes that will lead to too many collisions.
```

Comment:
```suggestion
  1. `hash` takes a value and produces a hash code with it. There is a lot of thought that goes into creating optimal hashing functions. It is up to you if you want to research to find a good hashing function, but don't dive too deep into the rabbit hole. Make sure to test that it's not producing too many similar hash codes. This will lead to frequent collisions when storing keys.
```

Reply:
Can you go over this again, please? I don't feel this present the idea that if we have a function that produces similar keys, then it will lead to frequent collisions. It seems you separated the concerns. Or we can re-arrange it to talk about frequent collisions earlier and that we need to avoid them.

Reply:
I like this change, and think it's better than mine :laughing: @01zulfi 

Do you have thoughts on it @Mclilzee 

Another thing I think we may want to specify here is: what types of things we can hash and use as keys in the hash map. With ruby hashes and JavaScript `Map`, you can use just about anything as a key (numbers, arrays, sets, objects). Under the hood, the hashing algorithm will generate different codes for things like `1` vs. `"1".

The other way would be like a JavaScript object implementation, where everything is just coerced to strings.
```js
const obj = {}
obj[1] = "value"; // stored a number key right

console.log(obj["1"]); // wrong
// logs "value"
```
Since the lesson only talked about how to deal with hashing strings, I think taking the 2nd approach would be best? And I also think it's probably a good idea to say that people will not be required to deal with hashing different types.

Reply:
It is still the same concept, the only thing that changes is the hash code input, the output is still a number. at this point, students should have enough experience to work around this.

We can add it as extra bonus to update the hash function so that it accepts any object and be able to generate meaningful hash code with it, Thoughts?

Reply:
IMO this could be a limitation as the project does seem quite lengthy as it stands. Something like:

> Hash maps could accommodate various data types for keys like numbers, strings, objects. But for this project, only handle keys of type strings.